packages/block-editor/src/components/global-styles/typography-panel.js

```diff
@@ -0,0 +1,470 @@
+/**
+ * WordPress dependencies
+ */
+import {
+	FontSizePicker,
+	__experimentalToolsPanel as ToolsPanel,
+	__experimentalToolsPanelItem as ToolsPanelItem,
+} from '@wordpress/components';
+import { __ } from '@wordpress/i18n';
+
+/**
+ * Internal dependencies
+ */
+import FontFamilyControl from '../font-family';
+import FontAppearanceControl from '../font-appearance-control';
+import LineHeightControl from '../line-height-control';
+import LetterSpacingControl from '../letter-spacing-control';
+import TextTransformControl from '../text-transform-control';
+import TextDecorationControl from '../text-decoration-control';
+import { getSupportedGlobalStylesPanels } from './hooks';
+import { getValueFromVariable } from './utils';
+
+export function useHasTypographyPanel( name, element, settings ) {
+	const hasFontFamily = useHasFontFamilyControl( name, element, settings );
+	const hasLineHeight = useHasLineHeightControl( name, element, settings );
+	const hasFontAppearance = useHasAppearanceControl(
+		name,
+		element,
+		settings
+	);
+	const hasLetterSpacing = useHasLetterSpacingControl(
+		name,
+		element,
+		settings
+	);
+	const hasTextTransform = useHasTextTransformControl(
```

Comment:
Each call of useHasYYYControl calls getSupportedGlobalStylesPanels which calculates all the supported panels. This results in the useHasTypographyPanel hook calculating all the supported panels 7 times. To improve performance, it would make sense to have the useHasYYYControl hooks already receive the supported panels as an argument, so that the calculation only needs to be done once.

Reply:
Ultimately, I'd like to refactor this to avoid the `name` and `element` props entirely here. In other words, I'd like to have a function that computes a `settings` prop by merging both global styles and block supports before calling this component. But I'd like to keep this in a separate PR, this one is already big enough.